File: Phase#1/Week#3/g.py
#!/usr/bin/python3

"""Run Time Error"""

if __name__ == '__main__':
    t = 1
    # t = int(input())

    for _ in range(t):
        def solve():
            s = list(input())
            costs = list(map(int, input().split()))
            n = len(s)

            def get_char(lower, upper=None):
                if not upper:
                    for i in range(26):
                        if costs[i] == lower:
                            return chr(i+97)
                else:
                    for i in range(26):
                        if costs[i] <= upper and costs[i] >= lower:
                            return chr(i+97)

            def get_cost(s):
                cnt = 0
                for i in range(1, n):
                    cnt += abs(costs[ord(s[i-1]) - 97]- costs[ord(s[i]) - 97])
                print(cnt)
                print(''.join(s))

            start = 0
            while s[start] == '?':
                start += 1
                if start >= n:
                    print(0)
                    print(''.join('a' for _ in range(n)))
                    break
            else:
                x = get_char(costs[ord(s[start]) - 97])
                s = [x for _ in range(start)] + s[start:]
                l = start
                while l < n:
                    while s[l+1] != '?':
                        l += 1
                        if l+1 >= n:
                            get_cost(s)
                            return

                    r = l+1
                    while s[r] == '?':
                        r += 1
                        if r >= n:
                            x = get_char(costs[ord(s[l]) - 97])
                            for i in range(l+1, n):
                                s[i] = x
                            # s = s[:l+1] + [get_char(costs[s[l]]) for _ in range(r-l-1)]
                            get_cost(s)
                            return
                    x = get_char(costs[ord(s[l]) - 97], costs[ord(s[r]) - 97])
                    for i in range(l+1, r):
                        s[i] = x
                    l = r
        solve()

File: Phase#1/Week#3/g.cpp
#include <iostream>
#include <vector>
#define endl "\n"
#define ll long long

using namespace std;

// wrong answer on test 18

void fastIO() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
}

char getChar(int lower, int upper=-1, vector<int> costs={}) {
    if (upper == -1) {
        upper = lower;
    }
    if (lower > upper) {
        swap(lower, upper);
    }
    for (int i = 0; i < 26; i++) {
        if (costs[i] >= lower && costs[i] <= upper) {
            return (char)('a' + i);
        }
    }
    return 'a';
}

void get_cost(string s, vector<int> costs) {
    int cost = 0;
    for (int i = 1; i < s.size(); i++) {
        cost += abs(costs[s[i] - 'a'] - costs[s[i - 1] - 'a']);
    }
    cout << cost << endl;
    cout << s << endl;
}

int main() {
    fastIO();
    string s;
    cin >> s;
    int n = s.size();
    vector<int> costs(26);
    for (int i = 0; i < 26; i++) {
        cin >> costs[i];
    }

    int start = 0;
    while (s[start] == '?') {
        start++;
        if (start == n) {
            cout << 0 << endl;
            string re = "";
            for (int i = 0; i < n; i++) {
                re += 'a';
            }
            cout << re << endl;
            return 0;
        }
    }

    char x;
    if (start > 0) {
        x = getChar(costs[s[start] - 'a'],  -1, costs);
        for (int i = start - 1; i >= 0; i--) {
            s[i] = x;
        }
    }

    int l = start;
    int r;
    while (l < n) {
        while (l+1 < n && s[l+1] != '?') {
            l++;
        }
        if (l == n) break;
        r = l + 1;
        while (r < n && s[r] == '?') {
            r++;
        }
        if (r == n) {
            x = getChar(costs[s[l] - 'a'], -1, costs);
            for (int i = l + 1; i < n; i++) {
                s[i] = x;
            }
            break;
        }
        x = getChar(costs[s[l] - 'a'], costs[s[r] - 'a'], costs);
        for (int i = l + 1; i < r; i++) {
            s[i] = x;
        }
        l = r;
    }
    get_cost(s, costs);
}
